feat: adding soulbound token (#393)
near/NEPs: neps/nep-0393.md

- [NEP-454](https://github.com/near/NEPs/pull/454) proposes royalties support for multi token contracts.

We propose that the SBT Standard will support the multi-token idea from the get go. This won't increase the complexity of the contract (in a traditional case, where one contract will only issue tokens of the single class, the `class` argument is simply ignored in the state, and in the functions it's required to be of a constant value, eg `1`) but will unify the interface.
It's up to the smart contract design how the token classes is managed. A smart contract can expose an admin function (example: `sbt_new_class() -> ClassId`) or hard code the pre-registered classes.

Finally, we require that each token ID is unique within the smart contract. This will allow us to query token only by token ID, without knowing it's class.

### SBT Registry

> Atomicity of _soul transfer_ in current NEAR runtime is not possible if the token balance is kept separately for each SBT smart contract. We need an additional contract: the `SBT Registry`, to provide atomic transfer of all user SBTs and efficient way to ban accounts in relation to the Ban event. The registry provides a balance book of all associated SBT tokens. So, each SBT is requested by a smart contract (issuer), however it's is created in the registry.

Reviewer: I don't think having centralised SBT Registries that actually hold these tokens is the right way to go. If this standard does want to adhere to the core ideas of NEP-171, then having contracts managing their own tokens is crucial.

Granted, it may solve the soul transfer question, however, I would suggest splitting these concerns. Having one on-chain entity only responsible for managing soul transfers, and a separate SBT contract, which knows how to communicate with this soul transfer governor.

Here is roughly what it would look like:
* You can issue a soul transfer against a governor. This would result in a soul transfer event noted in the governor. The governor can implement the various proposed validation steps.
* Each SBT contract has a list of trusted governors.
* Once a transfer is issued, you can apply the transfer on the SBT, which checks that the transfer was registered on the governor, in which case the transfer is completed

You can still burn accounts on governors, which the contracts can check themselves.

This allows the concept of recovery and soul transfer to evolve independently of the SBT Contracts themselves.
In your proposed solution, your SBTs are locked to the registry forever. Thus if a registry is vulnerable, all of those tokens are lost.

In the solution above, SBTs can pick which governors they support, thus not having to keep all state in sync in multiple registries. Instead, there is a single point of truth for the SBT, and multiple potential approaches to soul transfer and recovery. I'm sure you can see how this is more inline with the concept of SBTs.

Author: Registry could be fully permissionless without any admin feature. Again, we don't define in the standard how the registry governance should work. Hopefully there will be good market for it (as for example DEXes).

> Yes, except here one NFT collection is vulnerable. Not anyone who opted to use the registry.

Same for marketplaces, aggregators, escrows, exchanges....

Simple registry implementation will go through many reviews and we can do audits. Could be a template.

Reviewer: So what exactly happens if you do a soul transfer on one registry and not another? How are registries kept in sync with each other, for changes that occur on the registry?

> Simple registry implementation will go through many reviews and we can do audits

Yes, but with a governor you could just have an audited version deploy that others can use. You are decoupling SBT behavour and issuance and recovery/transfer.

Author: > https://github.com/near/NEPs/pull/393/#discussion_r1163938750

It depends on you. We can't restrict issuers to issue only one "copy" of SBT, and we can't restrict registries to be unique. In a same way, we can't restrict that NFT collections won't be copied, or issued twice (on the same chain or multiple chains).  
Ultimately, the credible ones will be used.

Reviewer: Yes, except that NFTs are designed to be a single source of truth. So a different contract is a different NFT. And you can clearly tell with branding and marketing which one is the real one, and which one is the copy.

The design here does not have a single source of truth. The same SBT on different registries are supposed to be the same. And there is no method of arbitration of which are supposed to be the "real" ones.

I feel like this is a fundamental flaw in this design, and it would be a mistake to go ahead as it is.

Today, the JS integer limit is `2^53-1 ~ 9e15`. Similarly, when minting 10'000 SBTs per second, it will take us 28'561 years to reach the limit. So, we don't need u128 nor the String type. However, if for some reason, we will need to get u64 support for JS, then we can always add another set of methods which will return String, so making it compatible with NFT standard (which is using `U128`, which is a string). Also, it's worth to note, that in 28'000 years JS (if it will still exists) will be completely different.

```rust
// TokenId and ClassId must be positive (0 is not a valid ID)
pub type TokenId = u64;
pub type ClassId = u64;

pub struct Token {
    pub token: TokenId,
    pub owner: AccountId,
    pub metadata: TokenMetadata,
}
```

The Soulbound Token follows the NFT [NEP-171](https://github.com/near/NEPs/blob/master/neps/nep-0171.md) interface, with few differences:

- token ID is `u64` (as discussed above).
- token class is `u64`, it's required when minting and it's part of the token metadata.

> - `TokenMetadata` doesn't have `title`, `description`, `media`, `media_hash`, `copies`, `extra`, `starts_at` nor `updated_at`. All that attributes except the `updated_at` can be part of the document stored at `reference`. `updated_at` can be tracked easily by indexers.

Reviewer: > `updated_at` can be tracked easily by indexers.

This is not correct. Method responses (Such as the response to `sbt()`) are not easy to track by indexers, as any TX to the contract may result in a change to any or all tokens.
The way NFTs solve this, using NEP-297, is by emitting a `contract_metadata_update`. Which is missing from this proposal.

Further reading: [NFT Events](https://nomicon.io/Standards/Tokens/NonFungibleToken/Event)

Reviewer: > keep track about updates, now how to listen on method response.

Indexers can listen to:
 - Method calls and their arguments, and responses
 - Events

Surely, with the wide range of options, many kind of method calls can cause token values to update. Which is not in the standard, and thus you have no reasonable expectation that indexers can implement each different registries.

Thus, you need to tell indexers that some metadata has changed. So you need the event that tells the indexers just that. They can then make a call and pull the updated metadata.

Author: @alexastrum actually there is `trait SBTContract` (the Token interface) in the reference implementation, but was not copied here. Thanks more mentioning it, I will add a section about it.

Reviewer: I raised it on this line, because you make a broad claim, without backing it up. The line 2 bellow just says you have events for it.
What I am highlighting is that your events are lacking. It does not emit for token updates. So there is no way to track that.
I'm happy to raise it again at the events if that is your concern. But the point stands. You do not have an update event, which lets indexers know a token has changed.

Reviewer 2: @robert-zaremba Maybe just remove that wording about tracking of `updated_at`, because it's currently not true according to the current NEP iteration: there is no methods nor events about metadata change (just expiration date can be tracked indirectly as a specific case of your API/events).

        limit: Option<u32>,
    ) -> Vec<TokenId>;

    /// Query SBT tokens by owner
    /// If `from_class` is not specified, then `from_class` should be assumed to be the first
    /// valid class id.
    /// Returns list of pairs: `(Contract address, list of token IDs)`.
    fn sbt_tokens_by_owner(
        &self,
        account: AccountId,
        ctr: Option<AccountId>,
        from_class: Option<u64>,
        limit: Option<u32>,
    ) -> Vec<(AccountId, Vec<TokenId>)>;

    /// checks if an `account` was banned by the registry.
    fn is_banned(&self, account: AccountId) -> bool;

    /*************
     * Transactions
     *************/

    /// Creates a new, unique token and assigns it to the `receiver`.
    /// `token_spec` is a vector of pairs: owner AccountId and TokenMetadata.
    /// Each TokenMetadata must have non zero `class`.
    /// Must be called by an SBT contract.
    /// Must emit `Mint` event.
    /// Must provide enough NEAR to cover registry storage cost.
    // #[payable]
    fn sbt_mint(&mut self, token_spec: Vec<(AccountId, TokenMetadata)>) -> Vec<TokenId>;

    /// sbt_recover reassigns all tokens from the old owner to a new owner,

>     /// and registers `old_owner` to a burned addresses registry.

Reviewer 3: Using the word `burn` for both accounts and tokens is confusing.

According to this spec no `Ban` events will be emitted here. But the only `Burn` event defined in this spec deals with tokens. Can we please rephrase?

Also see my concern regarding a potential privilege escalation attack vector, should this call `Ban` `old_owner` registry-wide.

    /// Must be called by an SBT contract.
    /// Must emit `Recover` event.
    /// Must provide enough NEAR to cover registry storage cost.
    /// Requires attaching enough tokens to cover the storage growth.
    // #[payable]
    fn sbt_recover(&mut self, from: AccountId, to: AccountId);

    /// sbt_renew will update the expire time of provided tokens.
    /// `expires_at` is a unix timestamp (in seconds).
    /// Must be called by an SBT contract.
    /// Must emit `Renew` event.
    fn sbt_renew(&mut self, tokens: Vec<TokenId>, expires_at: u64);

    /// Revokes SBT, could potentially burn it or update the expire time.
    /// Must be called by an SBT contract.
    /// Must emit one of `Revoke` or `Burn` event.
    /// Returns true if a token is a valid, active SBT. Otherwise returns false.
    fn sbt_revoke(&mut self, token: u64) -> bool;

    /// Transfers atomically all SBT tokens from one account to another account.
    /// The caller must be an SBT holder and the `to` must not be a banned account.
    /// Must emit `Revoke` event.
    // #[payable]
    fn sbt_soul_transfer(&mut self, to: AccountId) -> bool;
}
```

> SBT smart contracts can implement NFT query interface to make it compatible with NFT tools. Note, we use U64 type rather than U128.

Reviewer 3: I don't think we should suggest anyone implementing a fallback "query interface" without specifying explicitly what we expect this to be used for and by whom.

If we're not 100% sure, I'd suggest removing this section all-together, especially since the community decided to not emit NFT specific events.

  event: "mint" | "recover" | "renew" | "revoke" | "burn" | "ban" | "soul_transfer" ;
  data: Mint | Recover | Renew | Revoke | Burn | Ban[] | SoulTransfer;
}

/// An event minted by the Registry when new SBT is created.
type Mint {
  ctr: AccountId;    // SBT Contract recovering the tokens
  tokens: [[AccountId, u64[]]];  // list of pairs (token owner, TokenId[])
}

/// An event emitted when a recovery process succeeded to reassign SBT, usually due to account
/// access loss. This action is usually requested by the owner, but executed by an issuer,
/// and doesn't trigger Soul Transfer. Registry recovers all tokens assigned to `old_owner`,
/// hence we don't need to enumerate them.
/// Must be emitted by an SBT registry.
type Recover {
  ctr: AccountId         // SBT Contract recovering the tokens
  old_owner: AccountId;  // current holder of the SBT
  new_owner: AccountId;  // destination account.
}

/// An event emitted when a existing tokens are renewed.
/// Must be emitted by an SBT registry.
type Renew {
  ctr: AccountId;  // SBT Contract renewing the tokens
  tokens: u64[];  // list of token ids.
}

/// An event emitted when an existing tokens are revoked.

> /// Revoked tokens should not be listed in a wallet.

Reviewer 3: Can we be specific in the comments and state that the _Revoked tokens will continue to be listed by the registry_.

I assume this is the case here, unlike in the case of a Burn event.

Author: yes, thanks for suggestion.

/// in the future.
/// Must be emitted by an SBT registry.
type Ban {
  account: AccountId;
}

/// An event emitted when soul transfer is happening: all SBTs owned by `from` are transferred
/// to `to`, and the `from` account is banned (can't receive any new SBT).
/// Must be emitted by an SBT registry.
/// Registry MUST also emit `Ban` whenever the soul transfer happens.
type SoulTransfer {
  from: AccountId;
  to: AccountId;
}
```

Whenever a recovery is made in a way that an existing SBT is burned, the `Burn` event MUST be emitted. If `Revoke` burns token then `Burn` event MUST be emitted instead of `Revoke`.

### Example SBT Contract functions

Although the transaction functions below are not part of the SBT smart contract standard (depending on a use case, they may have different parameters), we present here an example interface for SBT issuance and we also provide a reference implementation.
These functions must relay calls to an SBT registry, which will emit appropriate events.
We recommend that the all functions related to an event will take an optional `memo: Option<String>` argument for accounting purposes.

```rust
trait SBT {

>     /// the function should overwrite `metadata.class = class`.

Reviewer 3: Why not just remove the `class` argument instead of having to explain that it overrides `metadata.class`?

    activate SBT1
    SBT1-)SBT_Registry: sbt_mint([[alice, metadata])
    SBT_Registry->>SBT_Registry: emit Mint(SBT_1_Contract, alice, [238])
    SBT_Registry-)SBT1: [238]
    deactivate SBT1

    Note over Alice,SBT_Registry: now Alice can query registry to check her SBT

    Alice-->>SBT_Registry: sbt(SBT_1_Contract, 238)
    SBT_Registry-->>Alice: {token: 238, owner: alice, metadata}

    Note over Alice,SBT_Registry: Alice participates in other community and applies for another SBT.<br/> She will get 2 SBTs. However she uses a different wallet: alice2.

    participant SBT2 as SBT_2 Contract
    actor Issuer2 as SBT_2 Issuer

    Issuer2->>SBT2: sbt_mint_multi([[alice2, metadata2], [alice2, metadata3]])
    activate SBT2
    SBT2-)SBT_Registry: sbt_mint([[alice2, metadata2], [alice2, metadata3]])
    SBT_Registry->>SBT_Registry: emit Mint(SBT_2_Contract, alice2, [7991, 17992])
    SBT_Registry-)SBT2: [7991, 1992]
    deactivate SBT2

    Alice-->>SBT_Registry: sbt(SBT_2_Contract, 7991)
    SBT_Registry-->>Alice: {token: 7991, owner: alice2, metadata: metadata2}

    Note over Alice,SBT_Registry: After a while Alice decides to merge her `alice2` account into `alice2`

>     Alice->>SBT_Registry: sbt_soul_transfer(alice) --accountId alice2

Reviewer: The notation here is not great. No other flows have specified which wallet Alice is using. I suggest having either two agents, as Alice1 and Alice2, or using a notation showing which of the accounts triggered which actions (both for `sbt_mint` and `sbt_mint_multi`).

Author: This is an example of one agent (Alice) who has 2 accounts and is clearly displayed by the diagram. 
(you should look at the diagram). If you are confused with `alice` being an account name, then I can rename it to `wallet1`.

Reviewer: No, I have looked at the diagram. And no, it is not clear. It may be clear to you, as you made it, and know the exact intent, but it is not serving its purpose of helping to explain the ideas better. Perhaps if the two SBT issuances were also triggered by Alice, then you can introduce this notation early, so it's more clear later.

Author: I have updated the diagram and split it into 3 diagrams. Hope it is more clear now. Thanks for asking about it.

Reviewer: Yes, it is much better. Thank you!

    Alice-->>+SBT_Registry: sbt_tokens_by_owner(alice2)
    SBT_Registry-->>-Alice: []

    Alice-->>+SBT_Registry: sbt_tokens_by_owner(alice)
    SBT_Registry-->>-Alice: [[SBT_1_Contract, [238]], [SBT_2_Contract, [7991, 7992]]]}
```

## Consequences

Being fully compatible with NFT standard is a desirable. However, given the requirements related to _soul transfer_ we didn't find an applaudable solution. Also we decided to use u64 as a Token ID, diverging further from the NFT NEP-171 standard.

Give that our requirements are much striker, we had to reconsider the level of compatibility with NEP-171 NFT.
There are many examples where NFT standards are poorly or improperly implemented, adding another standard with different functionality but equal naming will cause lots of problems and misclassifications between NFT and SBT.

### Positive

- Template and set of guidelines for creating SBT tokens.
- Ability to create SBT aggregators.

> - Ability to use SBT as a primitive to model non KYC identity, badges, certificates etc...

Author: But this requires a standard. Without a standard, we can't make interoperable ecosystem of such primitives. So, in my view, it is a positive consequence.

> - Ability to use SBT as a primitive to model non KYC identity, badges, certificates etc...

Reviewer: You can talk about the positive aspects of SBTs. But put them in a separate section. I would argue most people on NEAR agree that SBTs are a useful primitive. The question is, is this implementation of it right?

> - Ability to use SBT as a primitive to model non KYC identity, badges, certificates etc...

Author: OK, I will reformulate that sentence.

- SBT can be further used for "lego" protocols, like: Proof of Humanity (discussed for NDC Governance), undercollateralized lending, role based authentication systems, innovative economic and social applications...
- Standard recoverability mechanism.
- SBT are considered as a basic primitive for Decentralized Societies.
- new way to implement Sybil attack resistance.

### Neutral

- The API partially follows the NEP-171 (NFT) standard. The proposed design is to have native SBT API and make it possible for issuer contracts to support NFT based queries if needed (such contract will have a limitation of only issuing SBTs with one `ClassId` only).

### Negative

- New set of events to be handled by the indexer and wallets.

> - Complexity of integration with a registry: all SBT related transactions must go through Registry.

Reviewer: I find the lack of consideration for spam concerning. In this standard, there would inevitably be open registries. Using such registries it would be trivial to create thousands of tokens in an account, that they would have no recourse against.

Author: You are referring that an issuer issue SBTs to many wallets and doxxing account identity. This is a case of any public, permissionless, not fully private blockchian. There are multiple ways how someone can spam a account today:
* Create FT, block transfers (panic), and issue them to multiple accounts
* do the same with NFT
* do the same with any other primitive

Similarly, there are many ways how someone can use any media to doxx identity: making transactions with memo content about account identity, publishing a spreadsheet with associated personal data (IPFS, Tweets, Google spreadsheet....).

Registry has the `Burn` event, and it's up to the registry how to handle it. Accordingly, FT and NFT have a Burn event, but doesn't have standard function for that.

Token standard can't solve it.

Reviewer: > Create FT, block transfers (panic), and issue them to multiple accounts

This does not adhere to the NEP-141 standard, and wallets should rightfully ignore it.

> do the same with NFT

Same. Not standard compliant. Wallets should ignore them. Spam SBTs would be compliant to the standard.

Also, you are confusing two different problems. Doxxing, and spam. The two are very different concepts here. Yes, permanent doxxing is unavoidable in a public blockchain. Spam is not.

The token standard could solve this, if this was a priority, by having the user consent to their SBTs, otherwise hiding them/marking them as not valid.

Author: I don't understand why do you claim that spam NFT are not standard compliant and spam SBTs are.

Reviewer: If an NFT cannot be transferred, it does not adhere to the core of NEP-171. Thus, if wallets detect that an NFT cannot be transferred by their owner, they should rightfully start ignoring that NFT. TL;DR: An NFT that cannot be transferred is not an NEP-171 compliant NFT.
The same cannot be said about SBTs.

- Should we use NEP-171 Mint and NEP-171 Burn (instead of revoke) events? If the events will be emitted by registry, then we need new events to include the contract address. Since the native token ID is different in SBT and NFT, we should probably keep SBT events version. Also, with registry, it makes less sense to issue NEP-171 Mint and Burn.
- Confirm that only the registry emits the events.
- Decide if we need to keep memo in the events (it's already in the transaction).
- Should we keep the proposed multi-token approach?

## Changelog

### v1.0.0

#### Benefits

- Benefit
- Benefit

> #### Concerns

Reviewer 3: I left a few quick suggestions after catching up on reading the latest spec and the conversation.

Once I have time to dive into the reference implementation, I'll leave formal list of benefits and concerns.
Overall the spec looks good. Most of my concerns are around the registry restoration and revocation flows.

    /*************
     * Transactions
     *************/

    /// Creates a new, unique token and assigns it to the `receiver`.
    /// `token_spec` is a vector of pairs: owner AccountId and TokenMetadata.
    /// Each TokenMetadata must have non zero `class`.
    /// Must be called by an SBT contract.
    /// Must emit `Mint` event.
    /// Must provide enough NEAR to cover registry storage cost.
    // #[payable]
    fn sbt_mint(&mut self, token_spec: Vec<(AccountId, Vec<TokenMetadata>)>) -> Vec<TokenId>;

    /// sbt_recover reassigns all tokens from the old owner to a new owner,
    /// and registers `old_owner` to a burned addresses registry.
    /// Must be called by an SBT contract.
    /// Must emit `Recover` event.
    /// Must be called by an operator.
    /// Must provide enough NEAR to cover registry storage cost.
    /// Requires attaching enough tokens to cover the storage growth.
    // #[payable]
    fn sbt_recover(&mut self, from: AccountId, to: AccountId);

    /// sbt_renew will update the expire time of provided tokens.
    /// `expires_at` is a unix timestamp (in miliseconds).
    /// Must be called by an SBT contract.
    /// Must emit `Renew` event.

>     fn sbt_renew(&mut self, tokens: Vec<TokenId>, expires_at: u64);

Reviewer: This seems very use-case specific, and I don't think it belongs in the standard.

Author: Why? Remember that this is called by the issuer, not the user. So all the specificity is handled by the issuer contract, and the issuer will only be able to send a request to update the registry. Registry doesn't need to know that custom circumstances of the issuer.

Reviewer: You see, this again highlights the problem with the registry concept. If the registry doesn't support some feature, it's very difficult for the issuer to do it themselves, without becoming a registry themselves. Because I'd expect most SBTs wouldn't be renewable, but it cannot be an application specific interface, it has to be on the standard itself.

With any new primitive, it's not easy to predict the actual use cases. If the core interface doesn't cover that use case, it'll either result in new NEPs, or you end up with interface methods noone uses.

Reviewer 3: I wonder whether it makes sense to define `sbt_renew` as an (optional) extension to the core standard?

There are a few other capabilities in the SBT sample implementation that are not part of spec. So I don't see an issue in having a few capabilities like this one included in the spec. I will prefer however to keep the required capabilities to a minimum to make adoption easier.

    /// Revokes SBT, could potentially burn it or update the expire time.
    /// Must be called by an SBT contract.
    /// Must emit one of `Revoke` or `Burn` event.
    /// Returns true if a token is a valid, active SBT. Otherwise returns false.

>     fn sbt_revoke(&mut self, token: u64) -> bool;

Reviewer: It seems this adds very little value given that `sbt_burn` also exists. I suggest removing this from the standard, as it is completely redundant. If a token exists, but should not be displayed, it should just not exist.

Author: `sbt_revoke` and `sbt_burn` are very different. As documented, the former is called by the issuer, the latter by an owner. With `sbt_revoke` an issuer may revoke previously assigned SBT if it finds out that there was an abuse or a scam.

Reviewer: I still see this as two things that achieve the same thing. Why not just let the issuer burn tokens as well?

Also, I would suggest making it clearer which methods are meant to be called by issuers, and by the user (if you insist on sticking with the flawed registry concept).

Reviewer 3: This might be a need idea to semantically distinguish methods expected to be called by soul owners, from all the other API methods that are generally expected to be called by sbt emitter contracts.

Would having a prefix (or suffix) help (e.g `sbt_owner_burn`) or does a note in the method docs suffice?

```typescript
// only valid integer numbers (without rounding errors).
type u64 = number;

type Nep393Event {
  standard: "nep393";
  version: "1.0.0";
  event: "mint" | "recover" | "renew" | "revoke" | "burn" | "ban" | "soul_transfer" ;
  data: Mint | Recover | Renew | Revoke | Burn | Ban[] | SoulTransfer;
}

/// An event minted by the Registry when new SBT is created.
type Mint {
  ctr: AccountId;    // SBT Contract minting the tokens
  tokens: [[AccountId, u64[]]];  // list of pairs (token owner, TokenId[])
}

/// An event emitted when a recovery process succeeded to reassign SBTs, usually due to account
/// access loss. This action is usually requested by the owner, but executed by an issuer,
/// and doesn't trigger Soul Transfer. Registry reassigns all tokens assigned to `old_owner`
/// that were ONLY issued by the `ctr` SBT Contract (hence we don't need to enumerate the
/// token IDs).
/// Must be emitted by an SBT registry.
type Recover {
  ctr: AccountId         // SBT Contract recovering the tokens
  old_owner: AccountId;  // current holder of the SBT
  new_owner: AccountId;  // destination account.

> }

Author: No, Recover is triggered by issuer, and recovers only SBTs which can be recovered by issuer. It's like when you loose your ID: you need to go to your issuer to recover or re-issue it.

`SoulTransfer` is a user action, not an issuer action.

> }

Author: What do you mean? Standard defines how both Recovery and Soul Transfer works.

/// An event emitted when a existing tokens are renewed.
/// Must be emitted by an SBT registry.
type Renew {
  ctr: AccountId;  // SBT Contract renewing the tokens
  tokens: u64[];  // list of token ids.
}

/// An event emitted when an existing tokens are revoked.
/// Revoked tokens will continue to be listed by the registry but they should not be listed in
/// a wallet. See also `Burn` event.
/// Must be emitted by an SBT registry.
type Revoke {
  ctr: AccountId;  // SBT Contract revoking the tokens
  tokens: u64[];  // list of token ids.
}

/// An event emitted when an existing tokens are burned and removed from the registry.
/// Must be emitted by an SBT registry.
type Burn {
  ctr: AccountId;  // SBT Contract burning the tokens
  tokens: u64[];  // list of token ids.
}

/// An event emitted when the `account` is banned within the emitting registry.
/// Registry must add the `account` to a list of accounts that are not allowed to get any SBT
/// in the future.
/// Must be emitted by an SBT registry.
type Ban {
  account: AccountId;

> }

Reviewer: Would this ever be emitted without a `SoulTransfer`? If not, does having a separate event make sense?

Author: > Would a soul transfer issue two events, Ban and SoulTransfer

yes

> should a SoulTransfer also be treated as a Ban?

No. These events are not the same. Ban can be issued in other occasions, mentioned in the NEP.

Reviewer: But would a `SoulTransfer` ever not include a `Ban`? If so, why not enforce it by the protocol

Author: SoulTransfer emits `Ban` event:
> Whenever a soul transfer is triggered then the SBT registry emits `SoulTransfer` and `Ban` events

> More importantly, SBTs enable other applications of increasing ambition, such as community wallet recovery, Sybil-resistant governance, mechanisms for decentralization, and novel markets with decomposable, shared rights. We call this richer, pluralistic ecosystem “Decentralized Society” (DeSoc)—a co-determined sociality, where Souls and communities come together bottom-up, as emergent properties of each other to co-create plural network goods and intelligences, at a range of scales.

Creating a strong primitives is necessary to model new innovative systems and decentralized societies. Examples include reputation protocols, undercollateralized lending, one-person-one-vote, fair airdrops & ICOs, universal basic income, non KYC identity systems, Human DAOs and methods for Sybil attack resistance.

We propose an SBT standard to model protocols described above.

> _Verifiable Credentials_ (VC) could be seen as subset of SBT. However there is an important distinction: VC require set of claims and privacy protocols. It would make more sense to model VC with relation to W3 DID standard. SBT is different, it doesn't require a [resolver](https://www.w3.org/TR/did-core/#dfn-did-resolvers) nor [method](https://www.w3.org/TR/did-core/#dfn-did-methods) registry. For SBT, we need something more elastic than VC.

Reviewer 4: Lets add a link to W3 DID standard

Author: it's already there.

An issuer can provide an _sbt revocation_ in his contract (for example: when a related certificate or membership should be revoked, when an issuer finds out that there was an abuse or a scam, etc...). When doing so, the SBT registry MUST update token metadata and must emit the `Revoke` event. It's up to the registry to define revocation handling (either by burning a token, or changing expire date of it's metadata).

A registry can emit a `Ban` event without doing soul transfer. Handling it depends on the registry governance or registry use cases. One example is to use social governance to identify fake accounts (like bots) - in that case the registry should allow to emit `Ban` and block a scam soul and block future transfers.
NOTE: SBT smart contract (issuer) can have it's own list of blocked accounts or allowed only accounts.

### Token Class

SBT tokens can't be fractionized. Also, by definition there should be only one of a token per token class per user. Examples: user should not be able to receive few badges of the same class, or few proof of attendance to the same event.
However we identify a need for having few token classes in a single contract:

> - badges: one contract with multiple badge class (community lead, OG...);

Reviewer 4: I am not sure I understand it. On top of the class id there will be also badge id?

Author: I've updated the description to hopefully make it more clear.

- [NEP-454](https://github.com/near/NEPs/pull/454) proposes royalties support for multi token contracts.

We propose that the SBT Standard will support the multi-token idea from the get go. This won't increase the complexity of the contract (in a traditional case, where one contract will only issue tokens of the single class, the `class` argument is simply ignored in the state, and in the functions it's required to be of a constant value, eg `1`) but will unify the interface.
It's up to the smart contract design how the token classes is managed. A smart contract can expose an admin function (example: `sbt_new_class() -> ClassId`) or hard code the pre-registered classes.

Finally, we require that each token ID is unique within the smart contract. This will allow us to query token only by token ID, without knowing it's class.

### SBT Registry

> Atomicity of _soul transfer_ in current NEAR runtime is not possible if the token balance is kept separately for each SBT smart contract. We need an additional contract: the `SBT Registry`, to provide atomic transfer of all user SBTs and efficient way to ban accounts in relation to the Ban event. The registry provides a balance book of all associated SBT tokens. So, each SBT is requested by a smart contract (issuer), however it's is created in the registry.

Reviewer 4: ```suggestion
Atomicity of _soul transfer_ in current NEAR runtime is not possible if the token balance is kept separately for each SBT smart contract. We need an additional contract: the `SBT Registry`, to provide atomic transfer of all user SBTs and efficient way to ban accounts in relation to the Ban event. The registry provides a balance book of all associated SBT tokens. So, each SBT is requested by a smart contract (issuer), but created in the registry.
```

We can have multiple competing registries (with different purpose or different management scheme). Registries can utilize the same SBT contracts. An SBT smart contract, SHOULD opt-in to a registry. Registries may develop different opt-in mechanisms (they could differ by the approval mechanism, be fully permissioned etc..). One SBT smart contract can opt-in to:

- many registries: it MUST relay all state change functions to all registries.
- or to no registry: it MUST issue the SBT state change emits by itself.

If an SBT smart contract doesn't opt-in to any registry, then we should think about it as a single token registry, and it MUST strictly implement all SBT Registry query functions by itself: the contract address must be part of the arguments, and it must check that it equals to the deployed account address (`require!(ctr == env::current_account_id())`).

Moreover, a registry provides an efficient way to query multiple tokens for a single user. This will allow implementation of use cases such us:

- SBT based identities (main use case of the `i-am-human` protocol);
- SBT classes;
- decentralized societies.

#### Recovery within an SBT Registry

> SBT registry can define it's own mechanism to atomically recover all tokens related to one account and execute soul transfer to another account, without going one by one through each SBT issuer (sometimes that might be even not possible). Example mechanisms a Registry can implement to recovery mechanism:

Reviewer 4: ```suggestion
SBT registry can define it's own mechanism to atomically recover all tokens related to one account and execute soul transfer to another account, without going one by one through each SBT issuer (sometimes that might be even not possible). Example mechanisms a Registry can implement for the recovery:
```

- KYC based recovery
- Social recovery

![Social Recovery, Image via “Decentralized Society”](https://bankless.ghost.io/content/images/public/images/cdb1fc23-6179-44f0-9bfe-e5e5831492f7_1399x680.png)

SBT Registry based recovery is not part of this specification.

#### Minting authorization

A registry can limit which contracts can mint SBTs by implementing a custom issuer registration methods. Example: a simple access control list managed by a DAO.

#### Burning tokens

> Registry MAY expose a mechanism to allow an account to burn an unwanted token. The exact mechanism is not part of the standard and it will depend on the registry implementation. We only define a standard `Burn` event, which must be emitted each time a token is removed from existence. Some registries may not allow to let accounts burning their tokens in a spirit of preserving specific claims. Ultimately, NEAR is a public blockchain, and even if a token is burned, it's trace will be preserved.

Reviewer 4: ```suggestion
Registry MAY expose a mechanism to allow an account to burn an unwanted token. The exact mechanism is not part of the standard and it will depend on the registry implementation. We only define a standard `Burn` event, which must be emitted each time a token is removed from existence. Some registries may not allow to let accounts burn their tokens in a spirit of preserving specific claims. Ultimately, NEAR is a public blockchain, and even if a token is burned, it's trace will be preserved.
```

    /// not be included in the supply.
    fn sbt_supply(&self, issuer: AccountId) -> u64;

    /// Returns total amount of tokens of given class minted by `issuer`. See `sbt_supply` for
    /// information about revoked tokens.
    fn sbt_supply_by_class(&self, issuer: AccountId, class: ClassId) -> u64;

    /// Returns total supply of SBTs for a given owner. See `sbt_supply` for information about
    /// revoked tokens.
    /// If class is specified, returns only owner supply of the given class -- must be 0 or 1.
    fn sbt_supply_by_owner(
        &self,
        account: AccountId,
        issuer: AccountId,
        class: Option<ClassId>,
    ) -> u64;

    /// Query sbt tokens issued by a given contract.
    /// If `from_token` is not specified, then `from_token` should be assumed
    /// to be the first valid token id.
    fn sbt_tokens(
        &self,
        issuer: AccountId,
        from_token: Option<u64>,
        limit: Option<u32>,
    ) -> Vec<Token>;

    /// Query SBT tokens by owner
    /// If `from_class` is not specified, then `from_class` should be assumed to be the first

>     /// valid class id.

Reviewer 4: ```suggestion
    /// valid class id.
    /// `limit` specifies the upper limit of how many tokens we want to query
```

    /// Returns list of pairs: `(Contract address, list of token IDs)`.
    fn sbt_tokens_by_owner(
        &self,
        account: AccountId,
        issuer: Option<AccountId>,
        from_class: Option<u64>,
        limit: Option<u32>,
    ) -> Vec<(AccountId, Vec<OwnedToken>)>;

    /// checks if an `account` was banned by the registry.
    fn is_banned(&self, account: AccountId) -> bool;

    /*************
     * Transactions
     *************/

    /// Creates a new, unique token and assigns it to the `receiver`.
    /// `token_spec` is a vector of pairs: owner AccountId and TokenMetadata.
    /// Each TokenMetadata must have non zero `class`.

>     /// Must be called by an SBT contract.

Reviewer 4: ```suggestion
    /// Must be called by an SBT contract that has opted-in to the registry.
```

    /// Must emit `Mint` event.
    /// Must provide enough NEAR to cover registry storage cost.
    // #[payable]
    fn sbt_mint(&mut self, token_spec: Vec<(AccountId, Vec<TokenMetadata>)>) -> Vec<TokenId>;

    /// sbt_recover reassigns all tokens issued by the caller, from the old owner to a new owner.
    /// Adds `old_owner` to a banned accounts list.

>     /// Caller must be a valid issuer.

Reviewer 4: ```suggestion
```

    fn sbt_revoke(&mut self, token: Vec<TokenId>, burn: bool);
}
```

**Soul Transfer** MUST be part of the registry implementation. It is essential part of the standard, but the exact interface of the method is not part of the standard, because registries may adopt different mechanism and require different arguments. Moreover, big registries may require to implement the Soul Transfer operation in stages (and block the account in the meantime).
Smart contract must keep track of ongoing Soul Transfer operations. For example, in the first call, contract should lock the account, and do maximum X amount of transfers. If the list of to be transferred SBTs has not been exhausted, the contract should keep locking the account and remember the last transferred SBT. Subsequent calls by the same user will resume the operation until the list is exhausted.
Soult Transfer must emit the `SoulTransfer` event.

Example interface:

```rust
    /// Transfers atomically all SBT tokens from one account to another account.
    /// The caller must be an SBT holder and the `to` must not be a banned account.
    /// Returns amount of tokens transferred as well a boolean: `true` if the whole

>     /// process has finished, `false` when the process has not

Reviewer 4: ```suggestion
    /// process has finished, `false` if the process has not
```

  event: "mint" | "recover" | "renew" | "revoke" | "burn" | "ban" | "soul_transfer" ;
  data: Mint | Recover | Renew | Revoke | Burn | Ban[] | SoulTransfer;
}

/// An event minted by the Registry when new SBT is created.
type Mint {
  ctr: AccountId;    // SBT Contract minting the tokens
  tokens: [[AccountId, u64[]]];  // list of pairs (token owner, TokenId[])
}

/// An event emitted when a recovery process succeeded to reassign SBTs, usually due to account
/// access loss. This action is usually requested by the owner, but executed by an issuer,
/// and doesn't trigger Soul Transfer. Registry reassigns all tokens assigned to `old_owner`
/// that were ONLY issued by the `ctr` SBT Contract (hence we don't need to enumerate the
/// token IDs).
/// Must be emitted by an SBT registry.
type Recover {
  ctr: AccountId         // SBT Contract recovering the tokens
  old_owner: AccountId;  // current holder of the SBT
  new_owner: AccountId;  // destination account.
}

/// An event emitted when a existing tokens are renewed.
/// Must be emitted by an SBT registry.
type Renew {
  ctr: AccountId;  // SBT Contract renewing the tokens
  tokens: u64[];  // list of token ids.
}

> /// An event emitted when an existing tokens are revoked.

Reviewer 4: ```suggestion
/// An event emitted whe existing tokens are revoked.
```

/// Revoked tokens will continue to be listed by the registry but they should not be listed in
/// a wallet. See also `Burn` event.
/// Must be emitted by an SBT registry.
type Revoke {
  ctr: AccountId;  // SBT Contract revoking the tokens
  tokens: u64[];  // list of token ids.
}

> /// An event emitted when an existing tokens are burned and removed from the registry.

Reviewer 4: ```suggestion
/// An event emitted when existing tokens are burned and removed from the registry.
```

/// Must be emitted by an SBT registry.
type Burn {
  ctr: AccountId;  // SBT Contract burning the tokens
  tokens: u64[];  // list of token ids.
}

/// An event emitted when an account is banned within the emitting registry.
/// Registry must add the `account` to a list of accounts that are not allowed to get any SBT
/// in the future.
/// Must be emitted by an SBT registry.
type Ban = AccountId;

/// An event emitted when soul transfer is happening: all SBTs owned by `from` are transferred
/// to `to`, and the `from` account is banned (can't receive any new SBT).
/// Must be emitted by an SBT registry.
/// Registry MUST also emit `Ban` whenever the soul transfer happens.
type SoulTransfer {
  from: AccountId;
  to: AccountId;
}
```

Whenever a recovery is made in a way that an existing SBT is burned, the `Burn` event MUST be emitted. If `Revoke` burns token then `Burn` event MUST be emitted instead of `Revoke`.

### Example SBT Contract functions

> Although the transaction functions below are not part of the SBT smart contract standard (depending on a use case, they may have different parameters), we present here an example interface for SBT issuance and we also provide a reference implementation.

Reviewer 4: ```suggestion
Although the transaction functions below are not part of the SBT smart contract standard (depending on a use case, they may have different parameters), we present here an example of an interface for the SBT issuance and we also provide a reference implementation.
```

Author: I think the original text is more correct.

### Neutral

- The API partially follows the NEP-171 (NFT) standard. The proposed design is to have native SBT API and make it possible for issuer contracts to support NFT based queries if needed (such contract will have a limitation of only issuing SBTs with one `ClassId` only).

### Negative

- New set of events to be handled by the indexer and wallets.
- Complexity of integration with a registry: all SBT related transactions must go through Registry.

### Privacy Notes

> Blockchain-based systems are public by default. Any relationship that is recorded on-chain is immediately visible not just to the participants, but also to anyone in the entire world. Some privacy can be retained by having multiple pseudonyms: a family Soul, a medical Soul, a professional Soul, a political Soul each carrying different SBTs. But done naively, it could be very easy to correlate these Souls to each other.
> The consequences of this lack of privacy are serious. Indeed, without explicit measures taken to protect privacy, the “naive” vision of simply putting all SBTs on-chain may well make too much information public for many applications.

-- Decentralized Society

> There are multiple ways how a an identity can be doxxed using chain data. SBT, indeed provides more data about account.

Reviewer 4: ```suggestion
There are multiple ways how an identity can be doxxed using chain data. SBT, indeed provides more data about account.
```

Creating a strong primitives is necessary to model new innovative systems and decentralized societies. Examples include reputation protocols, undercollateralized lending, proof-of-personhood, proof-of-attendance, proof-of-skill, reputation protocols, one-person-one-vote, fair airdrops & ICOs, universal basic income, non KYC identity systems, Human DAOs and methods for Sybil attack resistance.

We propose an SBT standard to model protocols described above.

_Verifiable Credentials_ (VC) could be seen as subset of SBT. However there is an important distinction: VC require set of claims and privacy protocols. It would make more sense to model VC with relation to W3 DID standard. SBT is different, it doesn't require a [resolver](https://www.w3.org/TR/did-core/#dfn-did-resolvers) nor [method](https://www.w3.org/TR/did-core/#dfn-did-methods) registry. For SBT, we need something more elastic than VC.

## Specification

Main requirement for Soulbound tokens is to bound an account to a human. A **Soul** is an account with SBTs, which are used to define account identity. Moving SBTs from one account to another should be strictly limited to:

> - **recoverability** in case a user's private key is compromised due to extortion, loss, etc;
> - **soul transfer** - when a user needs to merge his accounts (e.g. they started with few different account but later decides to merge them to increase an account reputation). Soul transfer is different than a token transfer. The standard forbids a traditional token transfer functionality, where a user can transfer individual tokens. That being said, a registry can have extension functions for more advanced scenarios, which could require a governance mechanism.
> 
> This becomes especially important for proof-of-human stamps that can only be issued once per user. Few safeguards against misuse of SBT transfer (soul transfer or recovery) are contemplated:
> 
> 1. Users cannot recover an SBT by themselves. The issuer, a DAO or a smart contract (e.g. multisig) dedicated to manage the recovery should be assigned.
> 2. Whenever a soul transfer is triggered then the SBT registry emits `SoulTransfer` and `Ban` events. The latter signals that the account can't host nor receive any SBT in the future. It creates an inherit cost for such action: the account identity is burned.
> 3. `SoulTransfer` event can also trigger similar actions in other registries (specification for this is out of the scope of this NEP).
> 4. The recovery function is additional economical risk preventing account trading: user should always be able to recover his SBT, and move to another, not banned account.
> 
> SBT recover MUST not trigger `SoulTransfer` nor `Ban`: malicious issuer could compromise the system by faking the token recovery and take over all other SBTs from a user. Only the owner of the account can make a Soul Transfer transaction and merge 2 accounts they owns.
> 
> Soulbound tokens can have an _expire date_. This is useful for tokens which are related to real world certificates with expire time, or social mechanisms (e.g. community membership). Such tokens SHOULD have an option to be renewable. Examples include mandatory renewal with a frequency to check that the owner is still alive, or renew membership to a DAO that uses SBTs as membership gating.
> 
> An issuer can provide an _sbt revocation_ in his contract (for example: when a related certificate or membership should be revoked, when an issuer finds out that there was an abuse or a scam, etc...). Registry, when receiving `sbt_revoke` request from an issuer must always emit the `Revoke` event. Registry must only accept revoke requests from a valid issuer, and only revoke tokens from that issuer. If `burn=true` is set in the request, then the token should be burned and additionally `Burn` event must be emitted. Otherwise (when `burn=false`) the registry must update token metadata and set expire date in the past. Registry must not ban nor emit `Ban` event when revoking a contract. That would create an attack vector, when a malicious registry would thread the registry by banning accounts.
> 
> A registry can emit a `Ban` event without doing soul transfer. Handling it depends on the registry governance or registry use cases. One example is to use social governance to identify fake accounts (like bots) - in that case the registry should allow to emit `Ban` and block a scam soul and block future transfers.
> NOTE: SBT smart contract (issuer) can have it's own list of blocked accounts or allowed only accounts.
> 

Reviewer 2: I feel like it should be structured like this (definitions to definitions, constraints to constraints).

```suggestion
1. **soul transfer** when a user needs to merge his accounts (e.g. they started with few different account but later decides to merge them to increase an account reputation). Soul transfer is different than a token transfer. The standard forbids a traditional token transfer functionality, where a user can transfer individual tokens. That being said, a registry can have extension functions for more advanced scenarios, which could require a governance mechanism.
   * Whenever a soul transfer is triggered then the SBT registry emits `SoulTransfer` and `Ban` events. The latter signals that the account can't host nor receive any SBT in the future. It creates an inherit cost for such action: the account identity is burned.
   * `SoulTransfer` event can also trigger similar actions in other registries (specification for this is out of the scope of this NEP).
2. **recoverability** in case a user's private key is compromised due to extortion, loss, etc.
   * The recovery function is additional economical risk preventing account trading: user should always be able to recover his SBT, and move to another, not banned account.
3. A registry can emit a `Ban` event _without_ doing soul transfer.
   * Handling it depends on the registry governance or registry use cases. One example is to use social governance to identify fake accounts (like bots) - in that case the registry should allow to emit `Ban` and block a scam soul and block future transfers.
    NOTE: SBT smart contract (issuer) can have it's own list of blocked accounts or allowed only accounts.

This becomes especially important for proof-of-human stamps that can only be issued once per user.

Few safeguards against misuse of SBT transfer (soul transfer or recovery) are contemplated:

1. Users cannot recover an SBT by themselves. The issuer, a DAO or a smart contract (e.g. multisig) dedicated to manage the recovery should be assigned.
2. SBT recover MUST not trigger `SoulTransfer` nor `Ban`: malicious issuer could compromise the system by faking the token recovery and take over all other SBTs from a user. Only the owner of the account can make a Soul Transfer transaction and merge 2 accounts they owns.
3. Soulbound tokens can have an _expire date_. This is useful for tokens which are related to real world certificates with expire time, or social mechanisms (e.g. community membership). Such tokens SHOULD have an option to be renewable. Examples include mandatory renewal with a frequency to check that the owner is still alive, or renew membership to a DAO that uses SBTs as membership gating.
4. An issuer can provide an _sbt revocation_ in his contract (for example: when a related certificate or membership should be revoked, when an issuer finds out that there was an abuse or a scam, etc...). Registry, when receiving `sbt_revoke` request from an issuer must always emit the `Revoke` event. Registry must only accept revoke requests from a valid issuer, and only revoke tokens from that issuer. If `burn=true` is set in the request, then the token should be burned and additionally `Burn` event must be emitted. Otherwise (when `burn=false`) the registry must update token metadata and set expire date in the past. Registry must not ban nor emit `Ban` event when revoking a contract. That would create an attack vector, when a malicious registry would thread the registry by banning accounts.

```

Reviewer 2: @robert-zaremba Let me try to explain my point, my last try before you reject :) My perception of your narration is that you mixing up _safeguards/constraints_ and _functional requirements_ across the paragraphs:

1. [Constraint/safeguard] Users cannot recover an SBT by themselves...
2. [Functional requirement about transferability] Whenever a soul transfer is triggered then the SBT registry emits `SoulTransfer` and `Ban` events...
3. [Out-of-scope note] `SoulTransfer` event can also trigger similar actions... (specification for this is out of the scope of this NEP).
4. [Functional requirement/constraint about recoverability] .... user should always be able to recover his SBT, and move to another, not banned account.

[Constraint of recoverability functional requirement] SBT recover MUST not trigger `SoulTransfer` nor `Ban`...
[Functional requirement] Soulbound tokens can have an _expire date_. This is useful for tokens which...
[Functional requirement] An issuer can provide an _sbt revocation_ in his contract...
[Functional requirement  A registry can emit a `Ban` event without doing soul transfer. Handling it depends on the registry governance or registry use cases....
[Note :) ] NOTE: SBT smart contract....

In my example I grouped "kinda functional requirements" under 2 groups: `transferability` and `recoverability`, and safeguards go in a separate list.

Reviewer 5: I've read both your proposals a few times over. I can see benefits of both structures. It boils down to style, as the content is the same in both proposals. Robert's argument makes sense, there are some features and safeguards that span both soultransfer and recovery. My vote to keep it as is for this reason and because a lot of people have already read, commented and familiarized themselves based on the original storyline.

```rust
// TokenId and ClassId must be positive (0 is not a valid ID)
pub type TokenId = u64;
pub type ClassId = u64;

pub struct Token {
    pub token: TokenId,
    pub owner: AccountId,
    pub metadata: TokenMetadata,
}
```

The Soulbound Token follows the NFT [NEP-171](https://github.com/near/NEPs/blob/master/neps/nep-0171.md) interface, with few differences:

- token ID is `u64` (as discussed above).
- token class is `u64`, it's required when minting and it's part of the token metadata.
- `TokenMetadata` doesn't have `title`, `description`, `media`, `media_hash`, `copies`, `extra`, `starts_at` nor `updated_at`. All that attributes except the `updated_at` can be part of the document stored at `reference`. `updated_at` can be tracked easily by indexers.
- We don't have traditional transferability.
- We propose to use more targeted events, to better reflect the event nature. Moreover events are emitted by the registry, so we need to include issuer contract address in the event.

> -- TODO add mention that events will trigger metadata change

Reviewer 2: Check this TODO.....

```rust
trait SBT {
    /// the function should overwrite `metadata.class = class`.
    /// Must provide enough NEAR to cover registry storage cost.
    // #[payable]
    fn sbt_mint(
        &mut self,
        account: AccountId,
        class: u64,
        metadata: TokenMetadata,
        memo: Option<String>,
    ) -> TokenId;

    /// Creates a new, unique token and assigns it to the `receiver`.
    /// `token_spec` is a vector of pairs: owner AccountId and TokenMetadata.
    /// Must provide enough NEAR to cover registry storage cost.
    // #[payable]
    fn sbt_mint_multi(
        &mut self,
        token_spec: Vec<(AccountId, TokenMetadata)>,
        memo: Option<String>,
    ) -> Vec<TokenId>;

    // #[payable]
    fn sbt_recover(&mut self, from: AccountId, to: AccountId, memo: Option<String>);

    fn sbt_renew(&mut self, tokens: Vec<TokenId>, expires_at: u64, memo: Option<String>);

    fn sbt_revoke(token: Vec<u64>, memo: Option<String>) -> bool;
}
```

## Reference Implementation

- Common [type definitions](https://github.com/alpha-fi/i-am-human/tree/master/contracts/sbt) (events, traits).
- https://github.com/alpha-fi/i-am-human/tree/master/contracts/soulbound

> ## Example Flows

Reviewer 2: @robert-zaremba I would argue that this section (diagrams) is so important for understanding of the NEP that it should be at the top of the NEP right in the Specification section.

Reviewer 6: @robert-zaremba It is hard for me to keep small bits in memory unless I visualize / structure them, so diagrams are really helpful. At the very least, I would reference them in the first section(s).

SBT Registry based recovery is not part of this specification.

#### Minting authorization

A registry can limit which contracts can mint SBTs by implementing a custom issuer registration methods. Example: a simple access control list managed by a DAO.

#### Burning tokens

Registry MAY expose a mechanism to allow an account to burn an unwanted token. The exact mechanism is not part of the standard and it will depend on the registry implementation. We only define a standard `Burn` event, which must be emitted each time a token is removed from existence. Some registries may forbid accounts to burn their tokens in order to preserve specific claims. Ultimately, NEAR is a public blockchain, and even if a token is burned, it's trace will be preserved.

#### Personal Identifiable Information

Issuers must not include any PII into any SBT.

### Smart contract interface

For the Token ID type we propose `u64` rather than `U128`. `u64` capacity is more than 1e19. If we will mint 10'000 SBTs per second, then it will take us 58'494'241 years to fill the capacity.

> Today, the JS integer limit is `2^53-1 ~ 9e15`. Similarly, when minting 10'000 SBTs per second, it will take us 28'561 years to reach the limit. So, we don't need u128 nor the String type. However, if for some reason, we will need to get u64 support for JS, then we can always add another set of methods which will return String, so making it compatible with NFT standard (which is using `U128`, which is a string). Also, it's worth to note, that in 28'000 years JS (if it will still exists) will be completely different.

Reviewer 6: You don't make the life of JS folks any simpler. Parsing numbers bigger than $2^{53}$ from JSON payload is painful, so if you say that it is legal to have such big numbers, you force JS folks to use custom JSON parsers as browser parsers will lose the precision on deserialization OR give up on correctness/compliance from the get-go. Why do you want to put developers into this hassle?

I would recommend sticking to a stringified u64, or don't even define what string should contain, or standardizing the usage of u53 / u52.

Reviewer 6: > I want to stick to number, rather than string.

@robert-zaremba Then I would standardize TokenId to be u53, so anything beyond that can be treated as not compliant with the spec rather than a bug in the implementation that would be painful to solve.

> $2^53$ is large enough to cover the usage for next hundreds of years.

This is only true if you assume that the TokenIds will be allocated incrementally. Someone may decide to allocate ranges as IPv4 did, and then suddenly Tokens with 2^60 will be legal and cause problems on JS side.

We recommend that all functions related to an event will take an optional `memo: Option<String>` argument for accounting purposes.

```rust
trait SBT {
    /// the function should overwrite `metadata.class = class`.
    /// Must provide enough NEAR to cover registry storage cost.
    // #[payable]
    fn sbt_mint(
        &mut self,
        account: AccountId,
        metadata: TokenMetadata,
        memo: Option<String>,
    ) -> TokenId;

    /// Creates a new, unique token and assigns it to the `receiver`.
    /// `token_spec` is a vector of pairs: owner AccountId and TokenMetadata.
    /// Must provide enough NEAR to cover registry storage cost.
    // #[payable]
    fn sbt_mint_multi(
        &mut self,
        token_spec: Vec<(AccountId, TokenMetadata)>,
        memo: Option<String>,
    ) -> Vec<TokenId>;

    // #[payable]
    fn sbt_recover(&mut self, from: AccountId, to: AccountId, memo: Option<String>);

    fn sbt_renew(&mut self, tokens: Vec<TokenId>, expires_at: u64, memo: Option<String>);

    fn sbt_revoke(token: Vec<u64>, memo: Option<String>) -> bool;
}
```

## Reference Implementation

> - Common [type definitions](https://github.com/https://github.com/near-ndc/i-am-human/tree/main/contracts/sbt) (events, traits).
> - [I Am Human](https://github.com/https://github.com/near-ndc/i-am-human) registry and issuers.

Reviewer 6: ```suggestion
- Common [type definitions](https://github.com/near-ndc/i-am-human/tree/main/contracts/sbt) (events, traits).
- [I Am Human](https://github.com/near-ndc/i-am-human) registry and issuers.
```

## Example Flows

In the diagram below we explain a Soul Transfer use case. Alice has two accounts: `alice1` and `alice2`. She is getting tokens from 2 SBT contracts which use the same registry. Later she decides to merge the accounts by doing Soul Transfer.

#### Whitelist a new issuer

An `SBT_1 Issuer` wants to mint tokens using the `SBT_Registry`. Each registry has some way to whitelist an issuer. For this example, let's say that the `SBT_Registry` has a DAO which votes on adding a new issuer:

```mermaid
sequenceDiagram
    actor Issuer1 as SBT_1 Issuer
    actor DAO

    participant SBT_Registry

    Note over Issuer1,DAO: Issuer1 connects with the DAO<br>to be whitelisted.

    Issuer1-->>DAO: request whitelist
    DAO->>SBT_Registry: whitelist(SBT_1 Issuer)
```

#### Mint tokens

Alice uses her `alice1` account to interact with `SBT_1 Issuer` and receives an SBT with token ID = 238:

```mermaid
sequenceDiagram
    actor Alice
    actor Issuer1 as SBT_1 Issuer

    participant SBT1 as SBT_1 Contract
    participant SBT_Registry

>     Issuer1->>SBT1: sbt_mint(alice1, class5, metadata)

Reviewer 2: @robert-zaremba You upgraded `sbt_mint` prototype, you can remove `class5` from here now...

We also need an efficient on way to make composed ownership queries (for example: check if an account owns SBT of class C1, C2 and C3 issued by issuer I1, I2 and I3 respectively) - this is needed to model emergent properties discussed above.

We introduce a **soul transfer**: an ability for user to move ALL SBT tokens from one account to another in a [semi atomic](#soul-transfer) way, while keeping the SBT bounded to the same _soul_. This happens when a user needs to merge his accounts (e.g. they started with few different account but later decides to merge them to increase an account reputation). Soul transfer is different than a token transfer. The standard forbids a traditional token transfer functionality, where a user can transfer individual tokens. That being said, a registry can have extension functions for more advanced scenarios, which could require a governance approval.

SBT standard separates the token issuer concept from the token registry in order to meet the requirements listed above.
In the following sections we discuss the functionality of an issuer and registry.

### SBT Registry

> Traditional Token model in NEAR blockchain assumes that each token has it's own balance book and implements the authorization and issuance mechanism in the same smart contract. Such model prevents atomic _soul transfer_ in the current NEAR runtime. When token balance is kept separately in each SBT smart contract, synchronizing transfer calls to all such contracts to assure atomicity is not possible at scale. We need an additional contract: the `SBT Registry`, to provide atomic transfer of all user SBTs and efficient way to ban accounts in relation to the Ban event discussed below.

Reviewer 2: "...: the `SBT Registry` providing atomic transfer..."

This, and efficient cross-issuer queries are the main reasons SBT standards separates that token registry and token issuance concerns.

Issuer is an entity which issues new tokens and potentially can update the tokens (for example execute renewal). All standard modification options are discussed in the sections below.

Registry is a smart contract, where issuers register the tokens. Registry provides a balance book of all associated SBTs. Registry must ensure that each issuer has it's own "sandbox" and issuers won't overwrite each other. A registry provides an efficient way to query multiple tokens for a single user. This will allow implementation of use cases such us:

- SBT based identities (main use case of the `i-am-human` protocol);
- SBT classes;
- decentralized societies.

```mermaid
graph TB
    Issuer1--uses--> Registry
    Issuer2--uses--> Registry
    Issuer3--uses--> Registry
```

> We can have multiple competing registries (with different purpose or different management scheme). An SBT issuer, SHOULD opt-in to a registry before being able to use registry. Registries may develop different opt-in mechanisms (they could differ by the approval mechanism, be fully permissioned etc..). One SBT smart contract can opt-in to:

Reviewer 2: "An SBT issue <no comma> SHOULD opt-in"

Reviewer 6: @robert-zaremba Drop the comma after "issuer"

    /// If `from_token` is not specified, then `from_token` should be assumed
    /// to be the first valid token id. If `with_expired` is set to `true` then all the tokens are returned
    /// including expired ones otherwise only non-expired tokens are returned.
    fn sbt_tokens(
        &self,
        issuer: AccountId,
        from_token: Option<u64>,
        limit: Option<u32>,
        with_expired: bool,
    ) -> Vec<Token>;

    /// Query SBT tokens by owner.
    /// `limit` specifies the upper limit of how many tokens we want to return.
    /// If `from_class` is not specified, then `from_class` should be assumed to be the first
    /// valid class id. If `with_expired` is set to `true` then all the tokens are returned
    /// including expired ones otherwise only non-expired tokens are returned.
    /// Returns list of pairs: `(Contract address, list of token IDs)`.
    fn sbt_tokens_by_owner(
        &self,
        account: AccountId,
        issuer: Option<AccountId>,
        from_class: Option<u64>,
        limit: Option<u32>,
        with_expired: bool,
    ) -> Vec<(AccountId, Vec<OwnedToken>)>;

    /// checks if an `account` was banned by the registry.

>     fn is_banned(&self, account: AccountId) -> bool;

Reviewer 6: This is odd naming if you even look at it without deep context. `<contract>::is_banned` does not communicate enough context on its own. I would recommend `sbt_is_owner_banned` or something along those lines

    /*************
     * Transactions
     *************/

    /// Creates a new, unique token and assigns it to the `receiver`.
    /// `token_spec` is a vector of pairs: owner AccountId and TokenMetadata.
    /// Each TokenMetadata must have non zero `class`.
    /// Must be called by an SBT contract.
    /// Must emit `Mint` event.
    /// Must provide enough NEAR to cover registry storage cost.
    // #[payable]
    fn sbt_mint(&mut self, token_spec: Vec<(AccountId, Vec<TokenMetadata>)>) -> Vec<TokenId>;

    /// sbt_recover reassigns all tokens issued by the caller, from the old owner to a new one.
    /// Adds `old_owner` to a banned accounts list.
    /// Must not add `old_owner` to a ban list.
    /// Must be called by a valid SBT issuer.
    /// Must emit `Recover` event.
    /// Requires attaching enough NEAR to cover the storage growth.
    // #[payable]

>     fn sbt_recover(&mut self, from: AccountId, to: AccountId) -> (u32, bool);

Reviewer 6: I totally missed the return value here in the initial review.

`[1, true]`. Go figure what it means, and it is also not extendable. I recommend returning an object with explicit field names and avoid boolean fields, instead, return enum `"status": "FAILED" | "IN_PROGRESS" | "SUCCEEDED"` (I don't know what that bool signals in the first place)

Reviewer 6: @robert-zaremba I am not suggesting nullability, I suggest explicitly named fields and values, so return:

```json
{
  "status": "NEEDS_FURTHER_RECOVERY",
  "number_of_sbts_left": 10
}
```

instead of:

```json
[10, true]
```

The current type is not self-descriptive.

    /// Must emit `Revoke` event.
    /// Must also emit `Burn` event if the SBT tokens are burned (removed).
    fn sbt_revoke(&mut self, tokens: Vec<TokenId>, burn: bool);

    /// Similar to `sbt_revoke`, but revokes all `owner`s tokens issued by the caller.
    fn sbt_revoke_by_owner(&mut self, owner: AccountId, burn: bool);
}
```

Example **Soul Transfer** interface:

```rust
    /// Transfers atomically all SBT tokens from one account to another account.
    /// The caller must be an SBT holder and the `to` must not be a banned account.
    /// Returns amount of tokens transferred and a boolean: `true` if the whole
    /// process has finished, `false` when the process has not
    /// finished and should be continued by a subsequent call.
    /// User must keeps calling `sbt_soul_transfer` until `true` is returned.
    /// Must emit `SoulTransfer` event only if the caller was in possession of at least one SBT
    /// (if caller doesn't have any tokens, the function won't transfer anything, ban the
    /// original owner, emit Ban, and NOT emit SoulTransfer).
    #[payable]
    fn sbt_soul_transfer(
        &mut self,
        recipient: AccountId,

>     ) -> (u32, bool);

Reviewer 6: Same comment about return value type as for `sbt_recover` above.

Soul Transfer must emit the `SoulTransfer` event.

Example: Alice has two accounts: `alice1` and `alice2` which she used to mint tokens (see [mint diagram](#minting)). She decides to merge the accounts by doing Soul Transfer.

```mermaid
sequenceDiagram
    actor Alice
    participant SBT_Registry

    Alice->>SBT_Registry: sbt_soul_transfer(alice1) --accountId alice2

    Alice-->>+SBT_Registry: sbt_tokens_by_owner(alice2)
    SBT_Registry-->>-Alice: []

    Alice-->>+SBT_Registry: sbt_tokens_by_owner(alice1)
    SBT_Registry-->>-Alice: [[SBT_1_Contract, [238]], [SBT_2_Contract, [7991, 7992]]]}
```

### Renewal

> Soulbound tokens can have an _expire date_. This is useful for tokens which are related to real world certificates with expire time, or social mechanisms (e.g. community membership). Such tokens SHOULD have an option to be renewable. Examples include mandatory renewal with a frequency to check that the owner is still alive, or renew membership to a DAO that uses SBTs as membership gating.

Reviewer 2: "It is useful for tokens which are related to the real-world certificates with an expiration time...."

> Registry defines `sbt_renew` method which allows issuers to update the token expire date.

Reviewer 2: "that allows issuers to update the token expiration date"...

## Consequences

Being fully compatible with NFT standard is a desirable. However, given the requirements related to _soul transfer_ we didn't find an applaudable solution. Also we decided to use u64 as a Token ID, diverging further from the NFT NEP-171 standard.

Give that our requirements are much striker, we had to reconsider the level of compatibility with NEP-171 NFT.

> There are many examples where NFT standards are poorly or improperly implemented, adding another standard with different functionality but equal naming will cause lots of problems and misclassifications between NFT and SBT.

Author: OK

> There are many examples where NFT standards are poorly or improperly implemented, adding another standard with different functionality but equal naming will cause lots of problems and misclassifications between NFT and SBT.

Author: Actually, this serves as an argument for that latter sentence. WYT? How about only removing _poorly_?

> There are many examples where NFT standards are poorly or improperly implemented, adding another standard with different functionality but equal naming will cause lots of problems and misclassifications between NFT and SBT.

Reviewer 6: At the very least, remove "poorly", alternatively, rephrase the sentence. There is no point in making it sound like a drama.

#### Concerns

| #   | Concern                                                                                                                                                                                                                          | Resolution | Status |
| --- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------- | ------ |

> | 1   | Should we Emit NEP-171 Mint and NEP-171 Burn by the SBT contract (in addition to SBT native events emitted by the registry)? If the events will be emitted by registry, then we need new events to include the contract address. | -          | open   |

Author: The table is outdated, some of the concerns were added here to trigger discussion. The up to date table is in the PR description. In fact it seams we have an agreement everywhere ;)

- Common [type definitions](https://github.com/near-ndc/i-am-human/tree/main/contracts/sbt) (events, traits).
- [I Am Human](https://github.com/near-ndc/i-am-human) registry and issuers.

## Consequences

Being fully compatible with NFT standard is a desirable. However, given the requirements related to _soul transfer_ we didn't find an applaudable solution. Also we decided to use u64 as a Token ID, diverging further from the NFT NEP-171 standard.

Given that our requirements are much striker, we had to reconsider the level of compatibility with NEP-171 NFT.

> There are many examples where NFT standards are improperly implemented, adding another standard with different functionality but equal naming will cause lots of problems and misclassifications between NFT and SBT.

Reviewer 2: Make it via dot not a comma like this: "...implemented. Adding another standard..." as I realized I misread the sentence thrice.
